File: test/csharp/SpecGate.CtscFixtures/Conformance/Structured/DefaultInput.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Structured;

[SpecEvent("Offset")]
public sealed class Offset
{
    [SpecEvent("dx")]
    public int Dx { get; set; }

    [SpecEvent("dy")]
    public int Dy { get; set; }
}

public static class DefaultInput
{
    [SpecOperation("scale", Spec = "fixture.default_input")]
    public static int Scale(
        [SpecInput("value")] int value,
        [SpecInput("factor")] int factor) =>
        value * factor;

    [SpecOperation("shift", Spec = "fixture.default_input")]
    public static int Shift(
        [SpecInput("base")] int @base,
        [SpecInput("by")] Offset offset) =>
        @base + offset.Dx + offset.Dy;
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/basic/default_input.rs
use specgate::{SpecEvent, spec_operation};

#[spec_operation("scale", spec = "fixture.default_input")]
pub fn scale(value: i32, factor: i32) -> i32 {
    value * factor
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.default_input")]
pub struct Offset {
    #[spec_event]
    pub dx: i32,
    #[spec_event]
    pub dy: i32,
}

#[spec_operation("shift", spec = "fixture.default_input")]
pub fn shift(base: i32, by: Offset) -> i32 {
    base + by.dx + by.dy
}

#[test]
fn historical_default_values_are_explicit_capture_inputs() {
    assert_eq!(scale(6, 2), 12);
    assert_eq!(shift(10, Offset { dx: 1, dy: 1 }), 12);
}
